.github/workflows/haskell.yml

```diff
@@ -15,35 +15,55 @@ jobs:
 
     steps:
     - name: Checkout code
-      uses: actions/checkout@v3
+      uses: actions/checkout@v4
 
     - name: Install required libraries
       run: >-
         sudo apt install
         z3
 
+    # Use preinstalled Stack.  Should Stack fail to be installed, use the setup action:
+    # - name: Setup Haskell
+    #   uses: haskell-actions/setup@v2
+    #   with:
+    #     enable-stack: true
+    #     stack-no-global: true
 
-    - name: Cache Stack build artifacts (user + project)
-      uses: actions/cache@v3
+    - name: Restore Stack build artifacts (user + project)
+      uses: actions/cache/restore@v4
+      id: cache
       with:
         path: |
           ~/.stack
           .stack-work
-        # best effort for cache: tie it to Stack resolver and package config
-        key: ${{ runner.os }}-stack-${{ hashFiles('stack.yaml.lock', 'package.yaml') }}
+        # Write a fresh cache each time
+        key: ${{ runner.os }}-stack-${{ github.sha }}
         restore-keys: |
           ${{ runner.os }}-stack
 
-    
-    - name: Setup Haskell
-      uses: actions/setup-haskell@v1
-  
-    - name: Install GHC using Stack
-      if: steps.ghc.outputs.cache-hit != 'true'
-      run: stack setup --install-ghc  
-          
-    - name: Build
-      run: stack build
+    - name: Setup Stack
+      run: stack setup
+
+    - name: Build dependencies
+      run: stack test --only-dependencies
+
+    # This step only consumes an extra minute without gains.
+    # - name: Build w/o tests
+    #   run: stack build
+
+    - name: Build w/ tests
+      run: stack test --no-run-tests
 
     - name: Run tests
       run: stack test
+
+    # Always cache dependencies.
+    # This makes esp. sense now that some tests are broken.
```

Comment:
Can you clarify this comment? I am not up-to-date on best practices.

I am working on fixing the tests, so once they are fixed, should we not cache dependencies?

Reply:
No, sure.  
It is just that the secret sauce `always()` will enable caching _even_ if a previous step failed (such as building or running).

Reply:
Ah, that makes a lot of sense, thanks.